store/tikv/lock_resolver.go

```diff
@@ -0,0 +1,229 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tikv
+
+import (
+	"container/list"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+)
+
+const resolvedCacheSize = 512
+
+// LockResolver resolves locks and also caches resolved txn status.
+type LockResolver struct {
+	store *tikvStore
+	mu    sync.RWMutex
+	// Cache resolved txns (FIFO, txn id -> txnStatus).
+	resolved       map[uint64]txnStatus
+	recentResolved *list.List
+	// Maximum txnID that guaranteed to be expired.
+	maxExpire uint64
+}
+
+// NewLockResolver creates a LockResolver.
+func NewLockResolver(store *tikvStore) *LockResolver {
+	return &LockResolver{
+		store:          store,
+		resolved:       make(map[uint64]txnStatus),
+		recentResolved: list.New(),
+	}
+}
+
+type txnStatus uint64
+
+func (s txnStatus) isCommitted() bool { return s > 0 }
+func (s txnStatus) commitTS() uint64  { return uint64(s) }
+
+// locks after 3000ms is considered unusual (the client created the lock might
+// be dead). Other client may cleanup this kind of lock.
+// For locks created recently, we will do backoff and retry.
+var lockTTL uint64 = 3000
+
+// Lock represents a lock from tikv server.
+type Lock struct {
+	Key     []byte
+	Primary []byte
+	TxnID   uint64
+}
+
+func (lr *LockResolver) saveResolved(txnID uint64, status txnStatus) {
+	lr.mu.Lock()
+	defer lr.mu.Unlock()
+
+	if _, ok := lr.resolved[txnID]; ok {
+		return
+	}
+	lr.resolved[txnID] = status
+	lr.recentResolved.PushBack(txnID)
+	if len(lr.resolved) > resolvedCacheSize {
+		front := lr.recentResolved.Front()
+		delete(lr.resolved, front.Value.(uint64))
+		lr.recentResolved.Remove(front)
+	}
+}
+
+func (lr *LockResolver) getResolved(txnID uint64) (txnStatus, bool) {
+	lr.mu.RLock()
+	defer lr.mu.RUnlock()
+
+	s, ok := lr.resolved[txnID]
+	return s, ok
+}
+
+func (lr *LockResolver) isExpire(bo *Backoffer, l *Lock) (bool, error) {
+	lr.mu.RLock()
+	maxExpire := lr.maxExpire
+	lr.mu.RUnlock()
+
+	if l.TxnID <= maxExpire {
+		return true, nil
+	}
+	expired, err := lr.store.checkTimestampExpiredWithRetry(bo, l.TxnID, lockTTL)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if expired {
+		lr.mu.Lock()
+		if l.TxnID > lr.maxExpire {
+			lr.maxExpire = l.TxnID
+		}
+		lr.mu.Unlock()
+	}
+	return expired, nil
+}
+
+// ResolveLocks tries to resolve Locks. If returned `ok` is false, there
+// are some locks not expired, caller need to sleep then retry later.
+func (lr *LockResolver) ResolveLocks(bo *Backoffer, locks []*Lock) (ok bool, err error) {
+	var expiredLocks []*Lock
+	for _, l := range locks {
+		isExpired, err := lr.isExpire(bo, l)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+		if isExpired {
+			expiredLocks = append(expiredLocks, l)
+		}
+	}
+
```

Comment:
In most cases, there are no expiredLocks, here we can return early to avoid creating 'cleanTxns' map.